gtsam/hybrid/HybridFactorGraph.h

```diff
@@ -14,58 +14,24 @@
 
 #include <gtsam/discrete/DiscreteFactorGraph.h>
 #include <gtsam/hybrid/DCFactorGraph.h>
-#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+#include <gtsam/hybrid/GaussianHybridFactorGraph.h>
+#include <gtsam/hybrid/HybridBayesNet.h>
 
 #include <string>
 
 namespace gtsam {
 
-// Forward declarations
-class GaussianMixture;
-class Dummy;
-class HybridFactorGraph;
-class HybridEliminationTree;
-class Ordering;
-
-/** Main elimination function for HybridFactorGraph */
-using sharedFactor = boost::shared_ptr<Factor>;
-GTSAM_EXPORT std::pair<GaussianMixture::shared_ptr, sharedFactor>
-EliminateHybrid(const HybridFactorGraph& factors, const Ordering& keys);
-
-template <>
-struct EliminationTraits<HybridFactorGraph> {
-  typedef Factor FactorType;
-  typedef HybridFactorGraph FactorGraphType;
-  typedef GaussianMixture ConditionalType;
-  typedef HybridBayesNet BayesNetType;
-  typedef HybridEliminationTree EliminationTreeType;
-  typedef HybridBayesNet BayesTreeType;
-  typedef HybridEliminationTree JunctionTreeType;
-
-  /// The function type that does a single elimination step on a variable.
-  static std::pair<GaussianMixture::shared_ptr, sharedFactor> DefaultEliminate(
-      const HybridFactorGraph& factors, const Ordering& ordering) {
-    return EliminateHybrid(factors, ordering);
-  }
-};
-
-class HybridFactorGraph : protected FactorGraph<Factor>,
-                          public EliminateableFactorGraph<HybridFactorGraph> {
+template <typename FG>
+class GTSAM_EXPORT HybridFactorGraph : protected FactorGraph<Factor> {
  public:
   using shared_ptr = boost::shared_ptr<HybridFactorGraph>;
   using Base = FactorGraph<Factor>;
 
  protected:
   // Separate internal factor graphs for different types of factors
-  NonlinearFactorGraph nonlinearGraph_;
+  FG factorGraph_;
   DiscreteFactorGraph discreteGraph_;
   DCFactorGraph dcGraph_;
```

Comment:
This could still contain nonlinear things of course :-( Should we template on Factor type instead ? NonlinearFactor | GaussianFactor? I'd say leave for now but something to discuss.

Reply:
I like the templated version